crates/containerd-shim-wasmedge/src/instance.rs

```diff
@@ -103,171 +135,170 @@ pub fn reset_stdio() {
-
-                debug!("started wasi instance with tid {}", tid);
-
-                let code = self.exit_code.clone();
-
-                let _ = thread::spawn(move || {
-                    let (lock, cvar) = &*code;
-                    let status = match pidfd.wait() {
-                        Ok(status) => status,
-                        Err(e) => {
-                            error!("error waiting for pid {}: {}", tid, e);
-                            cvar.notify_all();
-                            return;
-                        }
-                    };
-
-                    debug!("wasi instance exited with status {}", status.status);
-                    let mut ec = lock.lock().unwrap();
-                    *ec = Some((status.status, Utc::now()));
-                    drop(ec);
-                    cvar.notify_all();
-                });
-                Ok(tid)
-            }
-            exec::Context::Child => {
-                // child process
-
-                // TODO: How to get exit code?
-                // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust.
-                let _ret = match vm.run_func(Some("main"), "_start", params!()) {
-                    Ok(_) => std::process::exit(0),
-                    Err(_) => std::process::exit(137),
-                };
-            }
-        }
+
+        let syscall = create_syscall();
+
+        fs::create_dir_all(&self.rootdir)?;
+
+        let mut container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
+            .with_executor(vec![Box::new(WasmEdgeExecutor {
+                stdin: maybe_open_stdio(self.stdin.as_str()).context("could not open stdin")?,
+                stdout: maybe_open_stdio(self.stdout.as_str()).context("could not open stdout")?,
+                stderr: maybe_open_stdio(self.stderr.as_str()).context("could not open stderr")?,
+            })])?
+            .with_root_path(self.rootdir.clone())?
+            .as_init(&self.bundle)
+            .with_systemd(false)
+            .build()?;
+
+        let code = self.exit_code.clone();
+        let pid = container.pid().unwrap();
+        container.start()?;
+
+        thread::spawn(move || {
+            let (lock, cvar) = &*code;
+
+            let status = match waitid(WaitID::Pid(pid), WaitPidFlag::WEXITED) {
+                Ok(WaitStatus::Exited(_, status)) => status,
+                Ok(WaitStatus::Signaled(_, sig, _)) => sig as i32,
+                Ok(_) => 0,
+                Err(e) => {
+                    if e == Errno::ECHILD {
+                        0
+                    } else {
+                        panic!("waitpid failed: {}", e);
+                    }
+                }
+            } as u32;
+            let mut ec = lock.lock().unwrap();
+            *ec = Some((status, Utc::now()));
+            drop(ec);
+            cvar.notify_all();
+        });
+
+        Ok(0)
```

Comment:
The docs say that this value should be a unique ID:

https://github.com/containerd/runwasi/blob/fff59017205828e1835ea7f699b7a0c5e7b924be/crates/containerd-shim-wasm/src/sandbox/instance.rs#L105-L108

Reply:
@rumpl 
I have updated this branch to fix it! Thanks for your review.